File: sdk/javascript/src/types/cryptography.ts
/**
 * Key pair interface
 */
export interface KeyPair {
  publicKey: string;
  privateKey: string;
  algorithm: string;
  createdAt: Date;
  expiresAt?: Date;
}

/**
 * Encrypted data
 */
export interface EncryptedData {
  ciphertext: string;
  nonce: string;
  algorithm: string;
  publicKey: string;
  timestamp: Date;
}

/**
 * Key rotation parameters
 */
export interface KeyRotationParams {
  address: string;
  oldPublicKey: string;
  newPublicKey: string;
  signature: string;
  reason?: string;
}

/**
 * Encryption parameters
 */
export interface EncryptParams {
  data: string;
  publicKey: string;
  algorithm?: string;
}

/**
 * Decryption parameters
 */
export interface DecryptParams {
  encryptedData: EncryptedData;
  privateKey: string;
}

/**
 * Quantum-resistant key pair
 */
export interface QuantumKeyPair extends KeyPair {
  quantumSafe: boolean;
  latticeParams?: {
    dimension: number;
    modulus: number;
  };
}

/**
 * Key info
 */
export interface KeyInfo {
  address: string;
  publicKey: string;
  algorithm: string;
  createdAt: Date;
  rotatedAt?: Date;
  rotationCount: number;
  isQuantumSafe: boolean;
  status: KeyStatus;
}

/**
 * Key status enum
 */
export enum KeyStatus {
  ACTIVE = 0,
  ROTATED = 1,
  REVOKED = 2,
  EXPIRED = 3,
}

/**
 * Cryptography parameters
 */
export interface CryptographyParams {
  defaultAlgorithm: string;
  keyRotationInterval: number;
  quantumSafeEnabled: boolean;
  maxKeyAge: number;
  supportedAlgorithms: string[];
}

/**
 * Secure enclave configuration
 */
export interface SecureEnclaveConfig {
  enabled: boolean;
  provider: string;
  attestation: string;
  keyStorageLevel: string;
}

/**
 * Random number request
 */
export interface RandomRequest {
  length: number;
  encoding?: 'hex' | 'base64' | 'bytes';
}

/**
 * Random number response
 */
export interface RandomResponse {
  value: string;
  entropy: number;
  timestamp: Date;
}

File: sdk/python/aura/types/cryptography.py
"""Type definitions for Cryptography module."""

from dataclasses import dataclass
from typing import Optional, List, Dict
from datetime import datetime
from enum import Enum


class KeyType(Enum):
    """Key types."""

    ED25519 = "ed25519"
    SECP256K1 = "secp256k1"
    RSA = "rsa"
    QUANTUM_RESISTANT = "quantum_resistant"


class EncryptionAlgorithm(Enum):
    """Encryption algorithms."""

    AES256 = "aes256"
    CHACHA20 = "chacha20"
    QUANTUM = "quantum"


@dataclass
class KeyRotationParams:
    """Parameters for key rotation."""

    address: str
    new_public_key: str
    key_type: KeyType
    proof: str


@dataclass
class EncryptionParams:
    """Parameters for encryption."""

    data: str
    algorithm: EncryptionAlgorithm
    public_key: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None


@dataclass
class DecryptionParams:
    """Parameters for decryption."""

    encrypted_data: str
    algorithm: EncryptionAlgorithm
    private_key: str


@dataclass
class KeyPair:
    """Cryptographic key pair."""

    public_key: str
    private_key: str
    key_type: KeyType
    created_at: datetime
    expires_at: Optional[datetime] = None


@dataclass
class EncryptedData:
    """Encrypted data structure."""

    ciphertext: str
    algorithm: EncryptionAlgorithm
    nonce: Optional[str] = None
    tag: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None


@dataclass
class SignatureParams:
    """Digital signature parameters."""

    message: str
    private_key: str
    algorithm: str


@dataclass
class VerificationParams:
    """Signature verification parameters."""

    message: str
    signature: str
    public_key: str
    algorithm: str


@dataclass
class RandomParams:
    """Random number generation parameters."""

    length: int
    algorithm: str = "secure"
    seed: Optional[str] = None


@dataclass
class SecureEnclaveParams:
    """Secure enclave parameters."""

    operation: str
    data: str
    enclave_id: str
    attestation: Optional[str] = None
